src/v/kafka/server/handlers/fetch.cc

```diff
@@ -456,95 +449,138 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                                       &plan,
+                                       &bytes_left_in_plan](
+                                        const fetch_session_partition& fp) {
+            // if this is not an initial fetch we are allowed to skip
+            // partions that aleready have an error or we have enough data
+            if (!octx.initial_fetch) {
+                bool has_enough_data = !resp_it->empty()
+                                       && octx.over_min_bytes();
+
+                if (resp_it->has_error() || has_enough_data) {
+                    ++resp_it;
+                    return;
+                }
+            }
 
-              auto shard = octx.rctx.shards().shard_for(ntp);
-              if (!shard) {
-                  /**
-                   * no shard is found on current node, but topic exists in
-                   * cluster metadata, this mean that the partition was moved
-                   * but consumer has not updated its metadata yet. we return
-                   * not_leader_for_partition error to force metadata update.
-                   */
-                  resp_it->set(make_partition_response_error(
-                    fp.partition, error_code::not_leader_for_partition));
-                  ++resp_it;
-                  return;
-              }
+            /**
+             * if not authorized do not include into a plan
+             */
+            if (!octx.rctx.authorized(
+                  security::acl_operation::read,
+                  fp.topic_partition.get_topic())) {
+                resp_it->set(make_partition_response_error(
+                  fp.topic_partition.get_partition(),
+                  error_code::topic_authorization_failed));
+                ++resp_it;
+                return;
+            }
 
-              auto fetch_md = octx.rctx.get_fetch_metadata_cache().get(ntp);
-              auto max_bytes = std::min(
-                bytes_left_in_plan, size_t(fp.max_bytes));
-              /**
-               * If offset is greater, assume that fetch will read max_bytes
-               */
-              if (fetch_md && fetch_md->high_watermark > fp.fetch_offset) {
-                  bytes_left_in_plan -= max_bytes;
-              }
+            auto& tp = fp.topic_partition;
+
+            auto shard = octx.rctx.shards().shard_for(tp);
+            if (unlikely(!shard)) {
+                // there is given partition in topic metadata, return
+                // unknown_topic_or_partition error
+
+                /**
+                 * no shard is found on current node, but topic exists in
+                 * cluster metadata, this mean that the partition was
+                 * moved but consumer has not updated its metadata yet. we
+                 * return not_leader_for_partition error to force metadata
+                 * update.
+                 */
+                auto ec = octx.rctx.metadata_cache().contains(tp.as_ntp())
```

Comment:
This is a key change: we move the metadata_cache().contains check inside the !shard_for condition. The idea is that we always need to do the `shard_for`, so do that first: if the ntp does not exist _anywhere_, i.e., not in the metadata, it also won't be in the shard list, we can defer the md.contains() to the case where it is not in the shard list.

This is one fewer hash table lookups in the common (non-error) case.